File: adventofcode/2019/d1/d1.scala
import scala.io.Source

object d1 extends App {
    val masses = Source.fromFile("input.txt").getLines.map(_.toInt).toList

    def calcFuel(mass: Int) =
        mass / 3 - 2

    println(masses.map(calcFuel).sum)

    def calcAllFuel(mass: Int) = Iterator.iterate(calcFuel(mass))(calcFuel)

    println(masses.map(calcAllFuel(_).takeWhile(_ > 0).sum).sum)
}

File: adventofcode/2019/d1/d1.jl
masses = readlines("input.txt")

fuel(x)::Int = x ÷ 3 - 2
fuel(x::String)::Int = fuel(parse(Int64, x))

# p1
@show masses .|> fuel |> sum

# p2
function true_fuel(mass)::Int
    f = fuel(mass)
    f <= 0 ? 0 : f + true_fuel(f)
end

@show masses .|> true_fuel |> sum
